File: server/internal/mcpserver/projections_networth.go
package mcpserver

import (
	"context"

	"tallyo/internal/auth"
	"tallyo/internal/graph/model"
	"tallyo/internal/money"
	u "tallyo/internal/utils"
)

type leanHolding struct {
	AssetID   string      `json:"assetId"`
	AccountID string      `json:"accountId"`
	Quantity  *float64    `json:"quantity"`
	ValueUsd  money.Cents `json:"valueUSD"`
	Manual    bool        `json:"manual"`
}

func mapHolding(h *model.Holding) leanHolding {
	return leanHolding{
		AssetID:   h.AssetID.EncodedString(),
		AccountID: h.AccountID.EncodedString(),
		Quantity:  h.Quantity,
		ValueUsd:  h.ValueUsd,
		Manual:    h.Manual,
	}
}

type leanHoldingRollup struct {
	Asset               leanAsset   `json:"asset"`
	TotalQuantity       *float64    `json:"totalQuantity"`
	ValueUsd            money.Cents `json:"valueUSD"`
	PercentOfClassifier float64     `json:"percentOfClassifier"`
	// Omitted (not empty) without read:holdings — an empty list would read as
	// "held in zero accounts", which is not what a forbidden field means.
	Holdings []leanHolding `json:"holdings,omitempty"`
}

func mapHoldingRollups(ctx context.Context, rollups []*model.HoldingRollup) []leanHoldingRollup {
	includeHoldings := auth.HasScope(ctx, auth.ScopeReadHoldings)
	toHoldingRollup := func(r *model.HoldingRollup) leanHoldingRollup {
		rollup := leanHoldingRollup{
			Asset:               mapAsset(r.Asset),
			TotalQuantity:       r.TotalQuantity,
			ValueUsd:            r.ValueUsd,
			PercentOfClassifier: r.PercentOfClassifier,
		}
		if includeHoldings {
			rollup.Holdings = u.Map(r.Holdings, mapHolding)
		}
		return rollup
	}
	return u.Map(rollups, toHoldingRollup)
}

type leanClassifierBreakdown struct {
	Classifier      model.AssetClassifier `json:"classifier"`
	Label           string                `json:"label"`
	ValueUsd        money.Cents           `json:"valueUSD"`
	PercentOfAssets float64               `json:"percentOfAssets"`
	AssetCount      int32                 `json:"assetCount"`
	Holdings        []leanHoldingRollup   `json:"holdings"`
}

type leanLiabilityBreakdown struct {
	Category             model.LiabilityCategory `json:"category"`
	Label                string                  `json:"label"`
	ValueUsd             money.Cents             `json:"valueUSD"`
	PercentOfLiabilities float64                 `json:"percentOfLiabilities"`
	AccountCount         int32                   `json:"accountCount"`
	Accounts             []leanAccount           `json:"accounts"`
}

type leanNetWorthReport struct {
	AsOfDate              *model.Date               `json:"asOfDate,omitempty"`
	CurrentNetWorthUsd    money.Cents               `json:"currentNetWorthUSD"`
	CurrentAssetsUsd      money.Cents               `json:"currentAssetsUSD"`
	CurrentLiabilitiesUsd money.Cents               `json:"currentLiabilitiesUSD"`
	ClassifierBreakdown   []leanClassifierBreakdown `json:"classifierBreakdown"`
	LiabilityBreakdown    []leanLiabilityBreakdown  `json:"liabilityBreakdown"`
}

func mapNetWorthReport(ctx context.Context, report *model.NetWorthReport) *leanNetWorthReport {
	toClassifier := func(c *model.ClassifierBreakdown) leanClassifierBreakdown {
		return leanClassifierBreakdown{
			Classifier:      c.Classifier,
			Label:           c.Label,
			ValueUsd:        c.ValueUsd,
			PercentOfAssets: c.PercentOfAssets,
			AssetCount:      c.AssetCount,
			Holdings:        mapHoldingRollups(ctx, c.Holdings),
		}
	}
	toLiability := func(l *model.LiabilityBreakdown) leanLiabilityBreakdown {
		return leanLiabilityBreakdown{
			Category:             l.Category,
			Label:                l.Label,
			ValueUsd:             l.ValueUsd,
			PercentOfLiabilities: l.PercentOfLiabilities,
			AccountCount:         l.AccountCount,
			Accounts:             mapAccounts(l.Accounts),
		}
	}
	return &leanNetWorthReport{
		AsOfDate:              report.AsOfDate,
		CurrentNetWorthUsd:    report.CurrentNetWorthUsd,
		CurrentAssetsUsd:      report.CurrentAssetsUsd,
		CurrentLiabilitiesUsd: report.CurrentLiabilitiesUsd,
		ClassifierBreakdown:   u.Map(report.ClassifierBreakdown, toClassifier),
		LiabilityBreakdown:    u.Map(report.LiabilityBreakdown, toLiability),
	}
}

File: server-rs/src/mcpserver/projections_networth.rs
use serde::Serialize;

use super::projections::{LeanAccount, LeanAsset, map_accounts, map_asset};
use crate::{
    auth::{Identity, Scope},
    ids::Date,
    money::Cents,
    schema::{
        AssetClassifier, ClassifierBreakdown, Holding, HoldingRollup, LiabilityBreakdown, LiabilityCategory,
        NetWorthReport,
    },
};

#[derive(Debug, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub(super) struct LeanHolding {
    pub(super) asset_id: String,
    pub(super) account_id: String,
    pub(super) quantity: Option<f64>,
    #[serde(rename = "valueUSD")]
    pub(super) value_usd: Cents,
    pub(super) manual: bool,
}

fn map_holding(holding: Holding) -> LeanHolding {
    LeanHolding {
        asset_id: holding.asset_id.to_string(),
        account_id: holding.account_id.to_string(),
        quantity: holding.quantity,
        value_usd: holding.value_usd,
        manual: holding.manual,
    }
}

#[derive(Debug, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub(super) struct LeanHoldingRollup {
    pub(super) asset: LeanAsset,
    pub(super) total_quantity: Option<f64>,
    #[serde(rename = "valueUSD")]
    pub(super) value_usd: Cents,
    pub(super) percent_of_classifier: f64,
    // Omitted (not empty) without read:holdings: an empty list would read as "held in zero accounts".
    #[serde(skip_serializing_if = "Vec::is_empty")]
    pub(super) holdings: Vec<LeanHolding>,
}

pub(super) fn map_holding_rollups(identity: &Identity, rollups: Vec<HoldingRollup>) -> Vec<LeanHoldingRollup> {
    let include_holdings = identity.has_scope(Scope::ReadHoldings);
    rollups
        .into_iter()
        .map(|rollup| LeanHoldingRollup {
            asset: map_asset(rollup.asset),
            total_quantity: rollup.total_quantity,
            value_usd: rollup.value_usd,
            percent_of_classifier: rollup.percent_of_classifier,
            holdings: rollup
                .holdings
                .filter(|_| include_holdings)
                .unwrap_or_default()
                .into_iter()
                .map(map_holding)
                .collect(),
        })
        .collect()
}

#[derive(Debug, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub(super) struct LeanClassifierBreakdown {
    pub(super) classifier: AssetClassifier,
    pub(super) label: String,
    #[serde(rename = "valueUSD")]
    pub(super) value_usd: Cents,
    pub(super) percent_of_assets: f64,
    pub(super) asset_count: i32,
    pub(super) holdings: Vec<LeanHoldingRollup>,
}

#[derive(Debug, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub(super) struct LeanLiabilityBreakdown {
    pub(super) category: LiabilityCategory,
    pub(super) label: String,
    #[serde(rename = "valueUSD")]
    pub(super) value_usd: Cents,
    pub(super) percent_of_liabilities: f64,
    pub(super) account_count: i32,
    pub(super) accounts: Vec<LeanAccount>,
}

#[derive(Debug, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub(super) struct LeanNetWorthReport {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub(super) as_of_date: Option<Date>,
    #[serde(rename = "currentNetWorthUSD")]
    pub(super) current_net_worth_usd: Cents,
    #[serde(rename = "currentAssetsUSD")]
    pub(super) current_assets_usd: Cents,
    #[serde(rename = "currentLiabilitiesUSD")]
    pub(super) current_liabilities_usd: Cents,
    pub(super) classifier_breakdown: Vec<LeanClassifierBreakdown>,
    pub(super) liability_breakdown: Vec<LeanLiabilityBreakdown>,
}

pub(super) fn map_net_worth_report(identity: &Identity, report: NetWorthReport) -> LeanNetWorthReport {
    let to_classifier = |breakdown: ClassifierBreakdown| LeanClassifierBreakdown {
        classifier: breakdown.classifier,
        label: breakdown.label,
        value_usd: breakdown.value_usd,
        percent_of_assets: breakdown.percent_of_assets,
        asset_count: breakdown.asset_count,
        holdings: map_holding_rollups(identity, breakdown.holdings),
    };
    let to_liability = |breakdown: LiabilityBreakdown| LeanLiabilityBreakdown {
        category: breakdown.category,
        label: breakdown.label,
        value_usd: breakdown.value_usd,
        percent_of_liabilities: breakdown.percent_of_liabilities,
        account_count: breakdown.account_count,
        accounts: map_accounts(breakdown.balances.into_iter().map(|balance| balance.account).collect()),
    };
    LeanNetWorthReport {
        as_of_date: report.as_of_date,
        current_net_worth_usd: report.current_net_worth_usd,
        current_assets_usd: report.current_assets_usd,
        current_liabilities_usd: report.current_liabilities_usd,
        classifier_breakdown: report.classifier_breakdown.into_iter().map(to_classifier).collect(),
        liability_breakdown: report.liability_breakdown.into_iter().map(to_liability).collect(),
    }
}
